src/session/content/background.rs

```diff
@@ -0,0 +1,397 @@
+
+            let style_manager = adw::StyleManager::default();
+            self.dark.set(style_manager.is_dark());
+
+            style_manager.connect_dark_notify(clone!(@weak obj => move |style_manager| {
+                let imp = obj.imp();
+                imp.dark.set(style_manager.is_dark());
+                imp.gradient_cache.take();
+            }));
+
+            let target = adw::CallbackAnimationTarget::new(clone!(@weak obj => move |progress| {
+                let imp = obj.imp();
+                imp.gradient_cache.take();
+                let progress = progress as f32;
+                if progress >= 1.0 {
+                    imp.progress.set(0.0);
+                    imp.phase.set((imp.phase.get() + 1) % 8);
+                } else {
+                    imp.progress.set(progress)
+                }
+                obj.queue_draw();
+            }));
+
+            let animation = adw::TimedAnimation::new(obj.as_ref(), 0.0, 1.0, 200, &target);
+            animation.set_easing(adw::Easing::EaseInOutQuad);
+
+            obj.connect_local("animate", true, move |_| {
+                let val = animation.value();
+                if val == 0.0 || val == 1.0 {
+                    animation.play()
+                }
+                None
+            });
+        }
+
+        fn signals() -> &'static [Signal] {
+            static SIGNALS: Lazy<Vec<Signal>> =
+                Lazy::new(|| vec![Signal::builder("animate").return_type::<()>().build()]);
+
+            SIGNALS.as_ref()
+        }
+    }
+
+    impl WidgetImpl for Background {
+        fn snapshot(&self, snapshot: &gtk::Snapshot) {
+            let widget = self.obj();
+            widget.ensure_shader();
+
+            let Some(Some([_, pattern_shader])) = &*self.shaders.borrow() else {
+                // fallback code
+                if let Some(child) = widget.child() {
+                    widget.snapshot_child(&child, snapshot);
+                }
+                return;
+            };
+
+            let width = widget.width() as f32;
+            let height = widget.height() as f32;
+
+            if width == 0.0 || height == 0.0 {
+                return;
+            }
+
+            let bounds = graphene::Rect::new(0.0, 0.0, width, height);
+
+            // background
+            if self.progress.get() == 0.0 {
+                let texture = {
+                    let size_changed = self.last_size.replace((width, height)) != (width, height);
+                    let cache = self.gradient_cache.borrow();
+
+                    match &*cache {
+                        Some(texture) if !size_changed => texture.clone(),
+                        _ => {
+                            drop(cache);
+
+                            let renderer = self.obj().native().unwrap().renderer();
+                            let texture = renderer
+                                .render_texture(self.gradient_shader_node(&bounds), Some(&bounds));
```

Comment:
I really think the gradient rendering should not be done here. It should be done every time the theme parameters change. In this case they never change, so you should only do it in the constructor. I suggest creating a method though, so that we can call it when we'll have full theme support. You can call it something like `render_freeform_gradient`.

Reply:
Oh, it should also be called here because you have to check if the size is change. But I think that moving all of that code in a method is still a good idea anyway. Then you can call if from here.